osu.Game/Rulesets/Mods/ModAlternate.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Graphics;
+using osu.Framework.Input.Bindings;
+using osu.Game.Beatmaps.Timing;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Rulesets.UI;
+
+namespace osu.Game.Rulesets.Mods
+{
+    public class ModAlternate : Mod
+    {
+        public override string Name => "Alternate";
+        public override string Acronym => "AL";
+        public override ModType Type => ModType.Fun;
+        public override double ScoreMultiplier => 1;
+        public override string Description => "Never use the same key twice!";
+        public override Type[] IncompatibleMods => new[] { typeof(ModAutoplay) };
+    }
+
+    public abstract class ModAlternate<THitObject, TAction> : ModAlternate, IApplicableToDrawableRuleset<THitObject>
+        where THitObject : HitObject
+        where TAction : struct
+    {
+        protected List<THitObject> HitObjects;
+        protected InputInterceptor Interceptor;
+        private readonly List<double> breakEndTimes = new List<double>();
+        private double nextreakTime;
+        private bool checkForBreaks = true;
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<THitObject> drawableRuleset)
+        {
+            HitObjects = drawableRuleset.Beatmap.HitObjects;
+            drawableRuleset.KeyBindingInputManager.Add(Interceptor = new InputInterceptor(this));
+
+            foreach (BreakPeriod period in drawableRuleset.Beatmap.Breaks)
+            {
+                var hitObject = HitObjects.First((h) => h.StartTime > period.EndTime);
+                var window = hitObject.HitWindows.WindowFor(HitResult.Miss);
+                breakEndTimes.Add(hitObject.StartTime - window);
+            }
+
+            try
+            {
+                nextreakTime = breakEndTimes.First();
+            }
+            catch
+            {
+                checkForBreaks = false;
+            }
+        }
+
+        protected abstract bool OnPressed(TAction action);
+
+        protected abstract bool OnReleased(TAction action);
+
+        protected abstract void ResetActionStates();
+
+        public void CheckForBreakEnd(double time)
+        {
+            if (checkForBreaks && time > nextreakTime)
+            {
+                ResetActionStates();
+
+                try
+                {
+                    nextreakTime = breakEndTimes.First((stamp) => stamp > time);
+                }
+                catch
+                {
+                    checkForBreaks = false;
+                }
```

Comment:
Same note about exceptions applies here.

Also, it would be nice if this could avoid enumerating an increasing number of elements of `breakEndTimes` every call, but I suppose it shouldn't have too many elements so maybe not really all that relevant here...